### DISCUSSION TITLE
Model binding with nullables doesn't work as an expect

### DISCUSSION THREAD

**Discussion by victore13** [laravel/framework#50029] [Q&A]
I have the problem using a component with Livewire, but after analyzing the code flow, I have discovered that where the unexpected occurs is in a native Laravel class.

## The problem
I have a class attribute of type MyModel|null.

In the mount() method of a Livewire component, I expect to receive a parameter of type MyModel|null and default to null. The problem occurs when in the component call, no argument is passed.

I would expect to receive the parameter as null since in the definition of the parameter's data type it could be null.

However, what I get is an object of type 'MyModel' with no attributes.

## Solutions?
What is the reason for this? If the parameter was not nullable, I understand that my attribute would be of type MyModel without attributes. But, however, I would expect to get a null.

## Example code
Call to the component without argument.
```blade
<livewire:my-component />
``` 

In the component class
```php
class MyComponent extends Component
{
    public ?MyModel $myModel = null;

    public function mount(?MyModel $myModel = null): void
    {
        dd($myModel); //Object of type MyModel without attributes
        $this->myModel = $myModel;
    }
}
``` 

I fact, made an example at [Wirebox](https://wirebox.app/b/x9l5k)

**Comment by henzeb** [comment]
I think this is not possible, due to the dependency injection from the container. This doesn't work in regular controllers also, it just resolves the model. 

You can work around it by either check if it's empty or remove the type from the definition, like so:

```php
class MyComponent extends Component
{
    public ?MyModel $myModel = null;

    public function mount($myModel = null): void
    {
        dd($myModel); // should be null now
        $this->myModel = $myModel;
    }
}
```

**Reply by victore13** [reply]
Thanks for responding, @henzeb!

I saw the alternative to my code, and it seemed like a good proposal, but I don't think it's a good practice not to define the datatype in variables/parameters.

I understand that injecting dependencies from the container in Laravel works like that, but I don't think that's the way to work when you've previously said that the parameter can be null.

Maybe an option would be to take into account the data type or the default value set before doing a model binding and, as it is in my case, instead of returning an empty model, return a null. Otherwise, if you want to get an empty model, it would be as simple as not setting the default value to null.

**Reply by henzeb** [reply]
I agree with you that it's not a good practice, I also find it a bit confusing. But it's how the dependency injector works. It's also how they recommend it in the docs. Besides, when you set a value that must be an instance of a certain model or null, the code will complain when its not what it expected either way.

**Reply by henzeb** [reply]
Now I think of it, if you really must, you can of course bind the model yourself, add something like `$app->bind(MyModel::class, fn()=>null)`. in your constructor.
